pvlib/iotools/pvgis.py

```diff
@@ -23,6 +23,346 @@
+        the error message in the response will be raised as an exception,
+        otherwise raise whatever ``HTTP/1.1`` error occurred
+
+    See Also
+    --------
+    pvlib.iotools.read_pvgis_hourly, pvlib.iotools.get_pvgis_tmy
+
+    References
+    ----------
+    .. [1] `PVGIS <https://ec.europa.eu/jrc/en/pvgis>`_
+    .. [2] `PVGIS Hourly Radiation
+       <https://ec.europa.eu/jrc/en/PVGIS/tools/hourly-radiation>`_
+    .. [3] `PVGIS Non-interactive service
+       <https://ec.europa.eu/jrc/en/PVGIS/docs/noninteractive>`_
+    .. [4] `PVGIS horizon profile tool
+       <https://ec.europa.eu/jrc/en/PVGIS/tools/horizon>`_
+    """  # noqa: E501
+    # use requests to format the query string by passing params dictionary
+    params = {'lat': latitude, 'lon': longitude, 'outputformat': outputformat,
+              'angle': surface_tilt, 'aspect': surface_azimuth,
+              'pvcalculation': int(pvcalculation),
+              'pvtechchoice': pvtechchoice, 'mountingplace': mountingplace,
+              'trackingtype': trackingtype, 'components': int(components),
+              'usehorizon': int(usehorizon),
+              'optimalangles': int(optimalangles),
+              'optimalinclination': int(optimalangles), 'loss': loss}
+    # pvgis only takes 0 for False, and 1 for True, not strings, also the
+    if userhorizon is not None:
+        params['userhorizon'] = ','.join(str(x) for x in userhorizon)
+    if raddatabase is not None:
+        params['raddatabase'] = raddatabase
+    if (start is not None) & (type(start) is int):
+        params['startyear'] = start
+    elif start is not None:
+        params['startyear'] = start if isinstance(start, int) else start.year
+    if end is not None:
+        params['endyear'] = end if isinstance(end, int) else end.year
+    if peakpower is not None:
+        params['peakpower'] = peakpower
+
+    # The url endpoint for hourly radiation is 'seriescalc'
+    res = requests.get(url + 'seriescalc', params=params, timeout=timeout)
+    # PVGIS returns really well formatted error messages in JSON for HTTP/1.1
+    # 400 BAD REQUEST so try to return that if possible, otherwise raise the
+    # HTTP/1.1 error caught by requests
+    if not res.ok:
+        try:
+            err_msg = res.json()
+        except Exception:
+            res.raise_for_status()
+        else:
+            raise requests.HTTPError(err_msg['message'])
+
+    if outputformat == 'json':
+        src = res.json()
+        return _parse_pvgis_hourly_json(src, map_variables=map_variables)
+    elif outputformat == 'csv':
+        with io.StringIO(res.content.decode('utf-8')) as src:
+            return _parse_pvgis_hourly_csv(src, map_variables=map_variables)
+    else:
+        # this line is never reached because if outputformat is not valid then
+        # the response is HTTP/1.1 400 BAD REQUEST which is handled earlier
+        raise ValueError('Invalid outputformat.')
```

Comment:
@wholmgren @kanderso-nrel Given that `read_pvgis_hourly` is able to take a file buffer, couldn't we just replace all these lines with:

    return read_pvgis_hourly(io.StringIO(res.text), pvgis_format=outputformat,
                             map_variables=map_variables)

Reply:
makes sense to me, no sense in duplicating the `outputformat` -> `_parse_pvgis_hourly_*` mapping.